File: ts/8.string-to-integer-atoi.ts
/*
 * @lc app=leetcode id=8 lang=typescript
 *
 * [8] String to Integer (atoi)
 */

// @lc code=start
function myAtoi(s: string): number {
  let sign = 1;
  let result = 0;
  let index = 0;

  let max = Math.pow(2, 31) - 1;
  let min = -Math.pow(2, 31);

  while (s[index] === " ") index++;

  if (s[index] === "+") {
    sign = 1;
    index++;
  } else if (s[index] === "-") {
    sign = -1;
    index++;
  }

  while (s[index] >= "0" && s[index] <= "9") {
    const digit = Number(s[index]);

    const isOverflow =
      result > Math.floor(max / 10) ||
      (result === Math.floor(max / 10) && digit > max % 10);

    if (isOverflow) return sign === 1 ? max : min;

    result = 10 * result + digit;
    index++;
  }

  return result * sign;
}
// @lc code=end

File: go/8.string-to-integer-atoi.go
/*
 * @lc app=leetcode id=8 lang=golang
 *
 * [8] String to Integer (atoi)
 */

package leetcode

import (
	"math"
	"strings"
)

// @lc code=start
func myAtoi(s string) int {
	s = strings.Trim(s, " ")

	if len(s) == 0 {
		return 0
	}

	var start int
	sign := 1

	if s[0] == '-' {
		sign = -1
		start = 1
	} else if s[0] == '+' {
		start = 1
	}

	var res int

	for i := start; i < len(s); i++ {
		char := s[i]

		if !(char >= '0' && char <= '9') {
			return res * sign
		}
		res = res*10 + int(char-'0')

		if sign*res >= math.MaxInt32 {
			return math.MaxInt32
		} else if sign*res <= math.MinInt32 {
			return math.MinInt32
		}
	}

	return sign * res
}

// @lc code=end
